src/block.rs

```diff
@@ -106,16 +93,16 @@ impl BlockBuilder {
     }
 
     #[must_use]
-    pub fn add(&mut self, key: &[u8], value: Option<&[u8]>, attrs: RowAttributes) -> bool {
-        assert!(!key.is_empty(), "key must not be empty");
-        let key_prefix_len = compute_prefix(&self.first_key, key);
-        let key_suffix = &key[key_prefix_len..];
+    pub fn add(&mut self, entry: RowEntry) -> bool {
+        assert!(!entry.key.is_empty(), "key must not be empty");
+        let key_prefix_len = compute_prefix(&self.first_key, &entry.key);
+        let key_suffix = entry.key.slice(key_prefix_len..);
 
         // If adding the key-value pair would exceed the block size limit, don't add it.
         // (Unless the block is empty, in which case, allow the block to exceed the limit.)
         if self.estimated_size()
                 + key_suffix.len()
-                + value.map(|v| v.len()).unwrap_or_default() // None takes no space (besides u32)
+                + entry.value.as_option().map(|v| v.len()).unwrap_or_default() // None takes no space (besides u32)
                 + SIZEOF_U16 * 3 // overlap key size + rest key size + offset size
                 + SIZEOF_U32 // value size
```

Comment:
Is this formula still accurate? Do we include a value size even if tombstone is set in `flags`?

Reply:
oh you're right, yes, we can not only change the value part here, since the key encoding has got more fields, we'd need to reflect the newly added fields into the estimated size here. let me fix it

Reply:
fixed in 7dc2e26, PTAL